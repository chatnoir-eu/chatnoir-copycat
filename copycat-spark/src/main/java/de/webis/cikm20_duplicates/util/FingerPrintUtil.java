package de.webis.cikm20_duplicates.util;

import java.io.Serializable;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import de.aitools.ir.fingerprinting.hashfunction.SimHash;
import de.aitools.ir.fingerprinting.representation.HashVector;
import de.aitools.ir.fingerprinting.representation.HashVectorSha3;
import de.aitools.ir.fingerprinting.representer.Hash;
import de.webis.cikm20_duplicates.spark.eval.SparkEvaluateSimHashFeatures;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.util.NGramms;
import info.debatty.java.lsh.MinHash;
import lombok.experimental.UtilityClass;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
@UtilityClass
public class FingerPrintUtil {
	public static interface Fingerprinter<T extends Comparable<T>> extends Serializable {
		public List<T> fingerprint(CollectionDocument doc);
		public List<T> fingerprint(List<String> features);
		
		public double similarity(List<T> a, List<T> b);
		public String fingerprinterName();
		
		public default double similarity(CollectionDocument a, CollectionDocument b) {
			List<T> fingerprintA = fingerprint(a);
			List<T> fingerprintB = fingerprint(b);
			
			return similarity(fingerprintA, fingerprintB);
		}
	}

	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> minHashFingerPrinting(long seed) {
		int dict_size = 1;
		int size = 12;
		MinHash minHash = new MinHash(size, dict_size, seed);
		
		return new Fingerprinter<Integer>() {
			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				return IntStream.of(minHash.signature(docToElementSet(doc))).boxed().collect(Collectors.toList());
			}
			
			@Override
			public List<Integer> fingerprint(List<String> features) {
				throw new RuntimeException();
			}

			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				int[] aArray = a.stream().mapToInt(i -> i).toArray();
				int[] bArray = b.stream().mapToInt(i -> i).toArray();
				
				return minHash.similarity(aArray, bArray);
			}
			
			public String fingerprinterName() {
				return "MinHashWithJavaHash";
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> simHashFingerPrinting(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				byte[] hash = hashVector(doc, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}

			@Override
			public List<Integer> fingerprint(List<String> features) {
				byte[] hash = hashVector(features, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}
			
			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				byte[] aArr = HashTransformationUtil.integersToHash(a);
				byte[] bArr = HashTransformationUtil.integersToHash(b);
				
				return 1.0 - (((double)Hash.getHammingDistance(aArr, bArr))/((double)aArr.length*8));
			}
			
			public String fingerprinterName() {
				return "64BitK3SimHashOneGramms";
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> productionFingerpringint(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				byte[] hash = hashVector(SparkEvaluateSimHashFeatures.theeAndFiveGramms(doc), bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}

			@Override
			public List<Integer> fingerprint(List<String> features) {
				byte[] hash = hashVector(features, bitsInSimHash);
				
				return HashTransformationUtil.hashToIntegers(hash, k);
			}
			
			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				byte[] aArr = HashTransformationUtil.integersToHash(a);
				byte[] bArr = HashTransformationUtil.integersToHash(b);
				
				return 1.0 - (((double)Hash.getHammingDistance(aArr, bArr))/((double)aArr.length*8));
			}
			
			public String fingerprinterName() {
				return "64BitK3SimHashThreeAndFiveGramms";
			}
		};
	}
	
	static byte[] hashVector(CollectionDocument doc, int bitsInSimHash) {
		return new SimHash().hash(internalHashVector(doc, bitsInSimHash)).getArray();
	}
	
	static byte[] hashVector(List<String> terms, int bitsInSimHash) {
		return new SimHash().hash(internalHashVector(terms, bitsInSimHash)).getArray();
	}
	
	public static HashVector internalHashVector(CollectionDocument doc, int bitsInSimHash) {
		List<String> terms = NGramms.tokenize(doc.getFullyCanonicalizedContent());
		return internalHashVector(terms, bitsInSimHash);
	}
	
	public static HashVector internalHashVector(List<String> terms, int bitsInSimHash) {
		return HashVectorSha3.toVector(terms, bitsInSimHash);
	}
	
	private static Set<Integer> docToElementSet(CollectionDocument doc) {
		return NGramms.build8Gramms(doc.getFullyCanonicalizedContent()).stream()
				.map(i -> i.getMd5Hash().hashCode())
				.collect(Collectors.toSet());
	}
}
