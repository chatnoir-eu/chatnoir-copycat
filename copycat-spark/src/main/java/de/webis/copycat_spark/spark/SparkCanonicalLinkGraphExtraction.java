package de.webis.copycat_spark.spark;

import java.io.IOException;
import java.io.Serializable;
import java.net.URL;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.InputFormat;
import org.apache.hadoop.mapred.InputSplit;
import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapred.RecordReader;
import org.apache.hadoop.mapred.Reporter;
import org.apache.hadoop.mapred.SequenceFileInputFormat;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaHadoopRDD;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import com.fasterxml.jackson.databind.ObjectMapper;

import de.webis.chatnoir2.mapfile_generator.inputformats.ClueWeb09InputFormat;
import de.webis.chatnoir2.mapfile_generator.warc.WarcRecord;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.index.transform.JsoupStringTransform;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;

public class SparkCanonicalLinkGraphExtraction {

//	@SuppressWarnings("unchecked")
//	public static void main(String[] args) {
////		String corpus = "cc-2017-04";
////		String sampleSize = "0.1";
////		String path = "/corpora/corpus-commoncrawl/CC-MAIN-2017-04-mapfile/data-r-*/data";
//		String corpus = "cc-2015-11";
//		String sampleSize = "0.1";
//		String path = "/corpora/corpus-commoncrawl/CC-MAIN-2015-11-mapfile/data-r-*/data";
//		
//		try (JavaSparkContext context = context()) {
//			Set<String> urlsToKeep = canonicalLinksToKeep(context.textFile("cikm2020/canonical-link-graph/" + corpus + "-canonical-urls-to-count-sample-" + sampleSize));
//			JavaHadoopRDD<Text, Text> rdd = (JavaHadoopRDD<Text, Text>) context.hadoopFile(path, SequenceFileInputFormat.class, Text.class, Text.class);	
//		
//			canonicalLinkedges(rdd, urlsToKeep).saveAsTextFile("cikm2020/canonical-link-graph/" + corpus + "-sample-" + sampleSize);
////			canonicalLinks(rdd).saveAsTextFile("cikm2020/canonical-link-graph/cc-2017-04-canonical-urls");
//		}
//	}

	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> tmp = context.textFile("s3a://corpus-copycat/document-representations/cw09");
			JavaPairRDD<LongWritable, WarcRecord> records = context.newAPIHadoopFile("", ClueWeb09InputFormat.class, LongWritable.class, WarcRecord.class, context.hadoopConfiguration());
			
			tmp.sample(false, 0.01).saveAsTextFile("tmp-maik-delete-me.jsonl");
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/canonical-link-graph");
	
		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> canonicalLinks(JavaPairRDD<Text, Text> input) {
		return input.map(i -> toCanonicalUrl(i._1().toString(), i._2().toString()))
				.filter(i -> i != null);
	}
	
	public static JavaRDD<String> canonicalLinkedges(JavaPairRDD<Text, Text> input, Set<String> urlsToKeep) {
		return input.map(i -> toVertex(i._1().toString(), i._2().toString(), urlsToKeep))
				.filter(i -> i != null)
				.map(i -> i.toString());
	}

	@SneakyThrows
	private static String toCanonicalUrl(String internalId, String json) {
		CanonicalLinkGraphEdge ret = toVertex(internalId, json, null);
		
		if(ret == null || ret.getCanonicalLink() == null) {
			return null;
		}
		
		return ret.getCanonicalLink().toString();
	}
	
	@SneakyThrows
	private static CanonicalLinkGraphEdge toVertex(String internalId, String json, Set<String> urlsToKeep) {
		// ignore large files
        if (json.getBytes().length > 1024 * 1024) {
            return null;
        }
        
        
        JSONObject inputJson  = new JSONObject(json);
        final JSONObject metadata = inputJson.getJSONObject("metadata");
        if (null == metadata) {
            throw new JSONException("Missing 'metadata'");
        }

        final JSONObject payload = inputJson.getJSONObject("payload");
        if (null == payload) {
            throw new JSONException("Missing 'payload'");
        }

        final String contentBody = payload.getString("body");
        final String contentEncoding    = payload.getString("encoding");

        if (null == contentEncoding || null == contentBody) {
            throw new JSONException("Missing one of 'payload/[encoding|body]'");
        }

        if (!contentEncoding.equals("plain")) {
            return null;
        }
        
        if(!"response".equals(metadata.getString("WARC-Type"))) {
        	return null;
        }

        String targetUri = metadata.getString("WARC-Target-URI");
        URL canonicalLink = extractCanonicalLinkOrNull(targetUri, contentBody);
        if(canonicalLink == null || (urlsToKeep != null && !urlsToKeep.contains(canonicalLink.toString()))) {
        	return null;
        }
        
        
        String id = internalId;
        if(metadata.has("WARC-TREC-ID")) {
        	id = metadata.getString("WARC-TREC-ID");
        }
        String timestamp = metadata.getString("WARC-Date");
        
        CollectionDocument ret = CollectionDocument.collectionDocument(new JsoupStringTransform().apply(contentBody), id);
        ret.setUrl(new URL(targetUri));
        
		return new CanonicalLinkGraphEdge(ret, canonicalLink, timestamp);
	}
	
	@SneakyThrows
	public static URL extractCanonicalLinkOrNull(String resolveFrom, String contentBody) {
		try {
			return extractCanonicalLinkOrNull(resolveFrom, Jsoup.parse(contentBody));
		} catch(Exception e) {
			return null;
		}
	}
	
	@SneakyThrows
	public static URL extractCanonicalLinkOrNull(String resolveFrom, Document doc) {
		try {
			Elements canonicals = doc.head().select("link[rel=\"canonical\"][href]");
			if(canonicals.size() == 0) {
				return null;
			}
	
			return new URL(new URL(resolveFrom), canonicals.get(0).attr("href"));
		} catch(Exception e) {
			return null;
		}
	}

	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class CanonicalLinkGraphEdge implements Serializable {
		private CollectionDocument doc;
		private URL canonicalLink;
		private String crawlingTimestamp;
		
		@SneakyThrows
		public String toString() {
			return new ObjectMapper().writeValueAsString(this);
		}
		
		@SneakyThrows
		public static CanonicalLinkGraphEdge fromString(String src) {
			return new ObjectMapper().readValue(src, CanonicalLinkGraphEdge.class);
		}
	}

	public static Set<String> canonicalLinksToKeep(JavaRDD<String> input) {
		List<String> ret = input.map(i -> urlFromUrlToCount(i))
				.filter(i -> i != null)
				.distinct().collect();
		
		return new HashSet<>(ret);
	}
	
	@SneakyThrows
	private static final String urlFromUrlToCount(String src) {
		return (String) new ObjectMapper().readValue(src, Map.class).get("url");
	}
}
