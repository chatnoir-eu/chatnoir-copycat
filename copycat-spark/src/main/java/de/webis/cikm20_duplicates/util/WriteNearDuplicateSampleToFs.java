package de.webis.cikm20_duplicates.util;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import de.webis.cikm20_duplicates.app.SampleNearDuplicates.NdSample;
import de.webis.cikm20_duplicates.spark.SparkEnrichRelevanceTransferPairs;
import de.webis.cikm20_duplicates.spark.SparkCalculateCanonicalLinkGraphEdgeLabels.CanonicalLinkGraphEdge2;
import de.webis.cikm20_duplicates.util.CollectionDocumentUtil.EsDocumentResolver;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import lombok.SneakyThrows;

public class WriteNearDuplicateSampleToFs {
	private static final String
	DIR = "src/main/resources/near-duplicate-sample/cc-2017-04/";
	
	//CREATED BY: de.webis.cikm20_duplicates.app.SampleNearDuplicatesIntegration
	private static final Path INPUT_PATH = Paths.get("src/main/resources/identified-near-duplicates-human-judgments/cw09.jsonl");
	
	
//	FIXME: Kann ich mittels S3 die kompletten near-duplicates berechnen, indem ich die Kandidaten final abteste?
//			Z.B. indem ich auf basis der near-duplicates geeignete blöcke bilde, die dann mit S3-Index getestet werden können?
//			Irgendwie in der Art: Die near-duplicate-Pairs geben mir irgendwie gruppen vor, mit denen ich nochmal nachberechnen kann.
//			Vielleicht über irgendwas transitives?
//			Mit neuen Berechnungen?
//			- für k=0 ist das machbar-> partitioniere Dokumente in den jeweiligen Block
//			- für k=1 muss man sich alle gruppen speichern, in die es verglichen werden muss?
//			  - Kann ich vielleicht die Idee anwenden, die in dem anderen Gedanken (bilde immer ab auf minimum) angewandt wurden?
//					  
//	- Nutze Elasticsearch fuer wahlfreien Zugriff auf Dokumente zur Pruefung
	public static void main(String[] args) {
		int count = 0;
		int wrong = 0;
		for(NdSample sample: samples(INPUT_PATH)) {
			CollectionDocument a = new EsDocumentResolver().loadCollectionDocument(sample.getFirstId());
			CollectionDocument b = new EsDocumentResolver().loadCollectionDocument(sample.getSecondId());
			double tmpScore = -1;
			if(a != null && b != null) {
				tmpScore = SparkEnrichRelevanceTransferPairs.s3Score(a,b);
			}
			
			System.out.println("hemming: " + sample.getHemmingDistance() +"; s3Score: " + sample.getS3Score() +"; mainContentS3Score: " + tmpScore);
			
			if(sample.getS3Score() < 0.82) {
//				System.out.println(sample.getS3Score() + " --> " + sample.getHemmingDistance());
			}
			
			int minHemming = 0;
			if(minHemming >= sample.getHemmingDistance()) {
				count++;
			}
			
			if(sample.getS3Score() < 0.82) {
//				System.out.println(sample.getFirstURL() + " vs " + sample.getSecondURL());
//				System.out.println(sample.getS3Score() + " --> " + sample.getHemmingDistance());
//				System.out.println();
				
//				System.out.println(sample);
				if(minHemming >= sample.getHemmingDistance()) {
					wrong++;
				}
			}
//			System.out.println(sample.getS3Score());
//			persistExample(sample)
		}
		
		System.out.println("---> " + (((double) wrong)/(double) count));
	}
	
	@SneakyThrows
	private static void persistExample(NdSample sample) {
		CollectionDocument src = sample.getFirstDocument();
		CollectionDocument target = sample.getSecondDocument();
		
		Path dir = Paths.get(DIR + src.getId() + "__to__" + target.getId());
		dir.toFile().mkdirs();
		
		Files.write(dir.resolve("metadata"), metadata(sample).getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("src-doc"), src.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("target-doc"), target.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
	}
	
	private static String metadata(NdSample sample) {
		return "First: "+ sample.getFirstURL() +"\n" + 
				"Second: "+ sample.getSecondId() +"\n" +
				"S3: " + sample.getS3Score() +"\n";
	}
	
	@SneakyThrows
	private static List<NdSample> samples(Path path) {
		List<String> lines = Files.readAllLines(path);
		List<NdSample> ret = new ArrayList<>();
		
		for(String line: lines) {
			ret.add(NdSample.fromString(line));
		}
		
		return ret;
	}
}
