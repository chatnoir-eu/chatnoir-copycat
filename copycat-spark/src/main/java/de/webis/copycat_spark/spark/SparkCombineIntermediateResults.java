package de.webis.copycat_spark.spark;

import java.util.Map;

import org.apache.commons.lang3.StringUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.SneakyThrows;

public class SparkCombineIntermediateResults {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> input = context.textFile("cikm2020/deduplication/near-duplicates/cw09-cw12");
			
			combineIntermediateResults(input, 5000)
				.saveAsTextFile("cikm2020/results/test-01");
		}
	}

	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/candidates");

		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> combineIntermediateResults(JavaRDD<String> inputRDD, int partitions) {
		return inputRDD
				.map(i -> toCsv(i))
				.filter(i -> i != null)
				.distinct()
				.sortBy(i -> i, Boolean.TRUE, partitions);
	}

	@SneakyThrows
	@SuppressWarnings("unchecked")
	private static String toCsv(String json) {
		if(StringUtils.isBlank(json)) {
			return null;
		}
		
		Map<String, Object> ret = new ObjectMapper().readValue(json, Map.class);
		String firstId = (String) ret.get("firstId");
		String secondId = (String) ret.get("secondId");
		Integer hemmingDistance = (Integer) ret.get("hemmingDistance");
		
		if(firstId == null|| secondId == null || hemmingDistance == null) {
			throw new IllegalArgumentException("Problem with input: '" + json + "'.");
		}
		
		return firstId + "," + secondId + "," + hemmingDistance;
	}
}
