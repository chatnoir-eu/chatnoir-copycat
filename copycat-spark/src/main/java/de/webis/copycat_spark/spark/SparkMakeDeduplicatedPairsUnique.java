package de.webis.copycat_spark.spark;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.SneakyThrows;
import scala.Tuple2;

public class SparkMakeDeduplicatedPairsUnique {
	@SneakyThrows
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			for(String corpus: new String[] {"cw09-cw12-cc-2015-11", "cc-2017-04", "cw09-cw12"}) {
				JavaRDD<String> nearDuplicatesWithoutExactDuplicates = context.textFile(inputPath(corpus));
				nearDuplicatesWithoutExactDuplicates.map(i -> nearDuplicatePair(i))
					.distinct()
					.saveAsTextFile("cikm2020/deduplication-final/64BitK3SimHashThreeAndFiveGramms/" + corpus + "-near-duplicates-without-exact-duplicates-csv-distinct");
				
			}
		}
	}
	
	public static JavaRDD<String> distinctNearDuplicatePairs(JavaRDD<String> nearDuplicateJSONL) {
		return nearDuplicateJSONL
				.mapToPair(i -> new Tuple2<>(nearDuplicatePair(i), i))
				.groupByKey().map(i -> combineRedundantNearDuplicateEntries(i));
	}
	
	private static String combineRedundantNearDuplicateEntries(Tuple2<String, Iterable<String>> i) {
		Iterator<String> values = i._2.iterator(); 
		String first = values.next();
		values.forEachRemaining(rem -> {
			if(rem != null && !rem.equals(first)) {
				throw new RuntimeException("Could not handle '" + first + "' vs '" + rem + "'.");
			}
		});
		
		return first;
	}

	@SneakyThrows
	@SuppressWarnings("unchecked")
	private static String nearDuplicatePair(String src) {
		Map<String, Object> parsed = new ObjectMapper().readValue(src, Map.class);
		String firstId = (String) parsed.get("firstId");
		String secondId = (String) parsed.get("secondId");
		List<String> ids = new ArrayList<>(Arrays.asList(firstId, secondId));
		Collections.sort(ids);
		
		return ids.get(0) + "," + ids.get(1);
	}
	
	private static String inputPath(String corpus) {
		if("cw09-cw12-cc-2015-11".equals(corpus)) {
			return "cikm2020/deduplication-final/64BitK3SimHashThreeAndFiveGramms/cw09-cw12-cc-2015-11-near-duplicates-without-exact-duplicates/part*/part*";
		} else if("cw09-cw12".equals(corpus)) {
			return "cikm2020/deduplication-final/64BitK3SimHashThreeAndFiveGramms/cw09-cw12-near-duplicates-without-exact-duplicates";
		} else if ("cc-2017-04".equals(corpus)) {
			return "cikm2020/deduplication-final/64BitK3SimHashThreeAndFiveGramms/cc-2017-04-near-duplicates-without-exact-duplicates/part*/part*";
		}
		
		throw new RuntimeException("Invalid corpus: " + corpus);
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/distinct-deduplication-pairs");

		return new JavaSparkContext(conf);
	}
}
