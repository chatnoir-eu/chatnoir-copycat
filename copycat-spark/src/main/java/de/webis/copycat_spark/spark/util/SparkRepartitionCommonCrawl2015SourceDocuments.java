package de.webis.copycat_spark.spark.util;

import org.apache.hadoop.io.compress.BZip2Codec;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class SparkRepartitionCommonCrawl2015SourceDocuments {
	
	private static final String[] CORPORA = new String[] {
			"cc-2015-11-part-0", "cc-2015-11-part-1", "cc-2015-11-part-2",
			"cc-2015-11-part-3", "cc-2015-11-part-4", "cc-2015-11-part-5",
			"cc-2015-11-part-6", "cc-2015-11-part-7", "cc-2015-11-part-8",
			"cc-2015-11-part-9"
	};
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			String corpus = "cc-2015-11";
			
			for(String c: CORPORA) {
				context.textFile("cikm2020/document-fingerprints-final/" + c + "-jsonl")
					.repartition(100)
					.saveAsTextFile("cikm2020/document-fingerprints-final/" + c + "-jsonl-repartioned");
			}
			
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints-final/" + corpus +"*-jsonl-repartioned");
			
			input.repartition(10000)
				.saveAsTextFile("cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2", BZip2Codec.class);
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");

		return new JavaSparkContext(conf);
	}
}
