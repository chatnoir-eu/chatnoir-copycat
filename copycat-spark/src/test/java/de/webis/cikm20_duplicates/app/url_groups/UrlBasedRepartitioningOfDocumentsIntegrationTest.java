package de.webis.cikm20_duplicates.app.url_groups;

import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;

import org.apache.spark.api.java.JavaRDD;
import org.approvaltests.Approvals;
import org.codehaus.jackson.map.ObjectMapper;
import org.junit.Assert;
import org.junit.Test;

import com.google.common.collect.ImmutableList;

import de.webis.cikm20_duplicates.spark.SparkIntegrationTestBase;
import lombok.SneakyThrows;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
public class UrlBasedRepartitioningOfDocumentsIntegrationTest extends SparkIntegrationTestBase {
	@Test
	public void approveRepartitioningOnSampleWith7Documents() {
		JavaRDD<String> input = asRDD(
			"{\"docId\":\"id-1\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-1.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[0,0,0,0],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-2\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-1.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[1,1,1,1],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-3\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-2.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[2,2,2,2],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-4\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-2.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[3,3,3,3],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-5\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-3.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[4,4,4,4],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-6\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-4.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[5,5,5,5],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-7\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-4.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[6,6,6,6],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}"
		);
		
		List<String> actual = mapToPartitions(input); 
		Approvals.verifyAsJson(actual);
	}

	@Test
	public void approveRepartitioningOnSampleWith7DocumentsInShuffledOrder() {
		JavaRDD<String> input = asRDD(
			"{\"docId\":\"id-1\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-1.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[0,0,0,0],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-2\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-1.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[1,1,1,1],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-3\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-2.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[2,2,2,2],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-4\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-2.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[3,3,3,3],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-5\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-3.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[4,4,4,4],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-6\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-4.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[5,5,5,5],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}",
			"{\"docId\":\"id-7\",\"url\":\"http://does-not-exist.com\",\"canonicalURL\":\"http://group-4.com\",\"fingerprints\":{\"64BitK3SimHashOneGramms\":[6,6,6,6],\"64BitK3SimHashThreeAndFiveGramms\":[-1,-1,-1,-1]}}"
		);
		
		List<String> actual = mapToPartitions(input); 
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void testDeserializationOfDocumentForUrlDeduplication() throws Exception {
		String input = "{\"docId\":\"my-id\", \"canonicalURL\": \"http://foo-bar.com\"}";
		DocumentForUrlDeduplication expected = new DocumentForUrlDeduplication(null, "my-id", new URL("http://foo-bar.com"));
		DocumentForUrlDeduplication actual = DocumentForUrlDeduplication.fromString(input);
		
		Assert.assertEquals(expected, actual);
	}
	
	private List<String> mapToPartitions(JavaRDD<String> input) {
		JavaRDD<String> ret = UrlBasedRepartitioningOfDocuments.repartitionDocumentsWithTheSameUrlOrCanonicalUrlToTheSamePart(input, 10);
		ret = ret.mapPartitions(i -> collectPartition(i));
		
		return ret.collect()
				.stream()
				.sorted()
				.collect(Collectors.toList());
	}

	@SneakyThrows
	private static Iterator<String> collectPartition(Iterator<String> i) {
		List<String> ret = ImmutableList.copyOf(i);
		return new ArrayList<>(Arrays.asList(new ObjectMapper().writeValueAsString(ret))).iterator();
	}
}
