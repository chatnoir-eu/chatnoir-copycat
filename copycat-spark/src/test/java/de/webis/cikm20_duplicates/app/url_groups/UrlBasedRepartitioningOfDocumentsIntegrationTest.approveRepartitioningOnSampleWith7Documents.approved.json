[
  "[\"{\\\"simHash64BitK3OneGramms\\\":[0,0,0,0],\\\"docId\\\":\\\"id-1\\\",\\\"canonicalURL\\\":\\\"http://group-1.com\\\"}\",\"{\\\"simHash64BitK3OneGramms\\\":[1,1,1,1],\\\"docId\\\":\\\"id-2\\\",\\\"canonicalURL\\\":\\\"http://group-1.com\\\"}\"]",
  "[\"{\\\"simHash64BitK3OneGramms\\\":[2,2,2,2],\\\"docId\\\":\\\"id-3\\\",\\\"canonicalURL\\\":\\\"http://group-2.com\\\"}\",\"{\\\"simHash64BitK3OneGramms\\\":[3,3,3,3],\\\"docId\\\":\\\"id-4\\\",\\\"canonicalURL\\\":\\\"http://group-2.com\\\"}\"]",
  "[\"{\\\"simHash64BitK3OneGramms\\\":[4,4,4,4],\\\"docId\\\":\\\"id-5\\\",\\\"canonicalURL\\\":\\\"http://group-3.com\\\"}\"]",
  "[\"{\\\"simHash64BitK3OneGramms\\\":[5,5,5,5],\\\"docId\\\":\\\"id-6\\\",\\\"canonicalURL\\\":\\\"http://group-4.com\\\"}\",\"{\\\"simHash64BitK3OneGramms\\\":[6,6,6,6],\\\"docId\\\":\\\"id-7\\\",\\\"canonicalURL\\\":\\\"http://group-4.com\\\"}\"]",
  "[]",
  "[]",
  "[]",
  "[]",
  "[]",
  "[]"
]